File: src/aizk/graph/insight.py
from loguru import logger
from patos import FrozenFlexModel
from pydantic import UUID5
from sqlmodel import select

from ..config import settings
from ..ontology import System
from ..serving.embed import Embedder
from ..serving.extract import LLM
from ..store import Entity, Fact
from ..store.engine import Session
from ..store.identity import User
from ..types import Scopes
from .dedupe import claim_entity, claim_fact
from .ids import entity_id, fact_id
from .models import InsightReport, Observation

# the single node every observation hangs off, one per user, so the derived insights form one
# small structural subgraph the recall fact lane already surfaces rather than a scattered set.
OBSERVATION_NODE = "graph observations"


def kept_observations(report: InsightReport) -> list[Observation]:
    """The observations that clear the significance gate, capped at the per-run write limit."""
    significant = [
        obs for obs in report.observations if obs.significance >= settings.insight_min_significance
    ]
    significant.sort(key=lambda obs: obs.significance, reverse=True)
    return significant[: settings.insight_max]


async def observation_already_claimed(session: Session, scopes: Scopes, identity: UUID5) -> bool:
    """Whether this scope already stakes an observes claim on this content id, ever."""
    claimed = (
        await session.exec(
            select(Fact.Claim.id)
            .where(
                Fact.Claim.content_id == identity,
                Fact.Claim.scopes == sorted(scopes),
            )
            .execution_options(**{settings.skip_live_gate: True})
        )
    ).first()
    return claimed is not None


async def write_observation(
    session: Session,
    scopes: Scopes,
    node_id: UUID5,
    obs: Observation,
    vector: list[float],
) -> bool:
    """Idempotently write one gated observation as an observes fact, returning whether it was
    new."""
    identity = fact_id(node_id, System.Relation.OBSERVES, None, obs.statement)
    if await observation_already_claimed(session, scopes, identity):
        return False
    await Fact.Content(
        id=identity,
        subject_id=node_id,
        object_id=None,
        predicate=System.Relation.OBSERVES,
        statement=obs.statement,
        embedding=vector,
    ).mint(session)
    await claim_fact(
        session,
        identity,
        settings.system_user_id,
        sorted(scopes),
        attributes={"significance": obs.significance},
    )
    return True


class InsightBuilder(FrozenFlexModel):
    """Derive and store significant observations in short database phases."""

    scopes: Scopes
    llm: LLM
    embed: Embedder

    async def grounding(self) -> list[str] | None:
        """The latest fact statements to reflect on, null when too few exist to ground on."""
        async with User.system(self.scopes) as session:
            statements = list(
                await session.exec(
                    Fact.Live.newest_statements(settings.insight_facts_k).where(
                        Fact.Live.predicate != System.Relation.OBSERVES
                    )
                )
            )
        if len(statements) < 2:
            logger.info("insight pass skipped for {}, too few facts to ground on", self.scopes)
            return None
        return statements

    async def store(self, kept: list[Observation], vectors: list[list[float]]) -> int:
        """Write gated observations as content-addressed, idempotent observes claims."""
        node_id = entity_id(OBSERVATION_NODE, System.Entity.OBSERVATION)
        async with User.system(self.scopes) as session:
            await Entity.Content(
                id=node_id,
                name=OBSERVATION_NODE,
                type=System.Entity.OBSERVATION,
            ).mint(session)
            await claim_entity(session, node_id, settings.system_user_id, sorted(self.scopes))
            written = sum(
                [
                    await write_observation(session, self.scopes, node_id, obs, vector)
                    for obs, vector in zip(kept, vectors, strict=True)
                ]
            )
        logger.info("insight pass wrote {} observations for {}", written, self.scopes)
        return written

    async def build(self) -> int:
        """Run the snapshot, model, embedding, and write phases."""
        grounding = await self.grounding()
        if grounding is None:
            return 0
        report = await self.llm.generate(
            settings.insight_system,
            "Facts:\n" + "\n".join(f"- {statement}" for statement in grounding),
            InsightReport,
        )
        kept = kept_observations(report)
        if not kept:
            logger.info(
                "insight pass wrote nothing for {}, no observation cleared the gate",
                self.scopes,
            )
            return 0
        vectors = await self.embed.embed(
            [observation.statement for observation in kept], mode="document"
        )
        return await self.store(kept, vectors)


async def derive_insights(
    llm: LLM,
    embed: Embedder,
    scopes: Scopes | None = None,
) -> int:
    """Derive observations from a user's graph and write the significant ones back."""
    builder = InsightBuilder(
        scopes=frozenset(scopes or (settings.system_user_id,)), llm=llm, embed=embed
    )
    return await builder.build()

File: archive/ts/src/lib/server/graph/insight.ts
// The reflective pass, the TS twin of graph/insight.py: read the latest fact statements as
// the system identity, ask the model for higher-level observations, gate them by
// significance, embed the keepers, and write each as a content-addressed observes fact
// hanging off one per-scope observation node, idempotent across runs.
import { sql } from 'drizzle-orm';

import { embed, structured } from '../serving';
import { settings } from '../settings';
import { entityId, factId } from './ids';
import { asSystem, scopeKey, uuidArray, type Reporter, type Tx } from './system';

const OBSERVATION = 'observation';
const OBSERVES = 'observes';

// The single node every observation hangs off, one per user, so the derived insights form
// one small structural subgraph the recall fact lane already surfaces.
const OBSERVATION_NODE = 'graph observations';

interface Observation {
	statement: string;
	significance: number;
}

interface InsightReport {
	observations: Observation[];
}

const INSIGHT_SCHEMA = {
	name: 'InsightReport',
	schema: {
		type: 'object',
		properties: {
			observations: {
				type: 'array',
				items: {
					type: 'object',
					properties: {
						statement: {
							type: 'string',
							description: 'a self-contained insight grounded only in the facts shown'
						},
						significance: {
							type: 'number',
							minimum: 0,
							maximum: 1,
							description: 'how much the insight adds beyond the facts, from 0 to 1'
						}
					},
					required: ['statement', 'significance'],
					additionalProperties: false
				}
			}
		},
		required: ['observations'],
		additionalProperties: false
	}
};

const report: Reporter<InsightReport> = (system, user) =>
	structured<InsightReport>(system, user, INSIGHT_SCHEMA);

// The observations that clear the significance gate, capped at the per-run write limit.
const keptObservations = (insight: InsightReport): Observation[] =>
	insight.observations
		.filter((obs) => obs.significance >= settings.insightMinSignificance)
		.sort((a, b) => b.significance - a.significance)
		.slice(0, settings.insightMax);

// Whether this scope already stakes an observes claim on this content id, ever, so the
// check reads past the live gate the recall lanes apply.
const alreadyClaimed = async (tx: Tx, key: string[], identity: string): Promise<boolean> => {
	const claimed = (await tx.execute(sql`
		select id from fact_claim
		where content_id = ${identity} and scopes = ${uuidArray(key)}::uuid[]
		limit 1
	`)) as unknown as { id: string }[];
	return claimed.length > 0;
};

// Idempotently write one gated observation as an observes fact, returning whether it was
// new. Content minting tolerates the deterministic-id race exactly like the Python mint().
const writeObservation = async (
	tx: Tx,
	key: string[],
	nodeId: string,
	obs: Observation,
	vector: string
): Promise<boolean> => {
	const identity = factId(OBSERVATION_NODE, OBSERVES, '', obs.statement);
	if (await alreadyClaimed(tx, key, identity)) return false;
	await tx.execute(sql`
		insert into fact_content (id, subject_id, object_id, predicate, statement, embedding)
		values (${identity}, ${nodeId}, null, ${OBSERVES}, ${obs.statement}, ${vector}::halfvec)
		on conflict do nothing
	`);
	await tx.execute(sql`
		insert into fact_claim (id, content_id, created_by, scopes, attributes)
		values (uuidv7(), ${identity}, ${settings.systemUserId}, ${uuidArray(key)}::uuid[],
			jsonb_build_object('significance', ${obs.significance}::float8))
		on conflict (content_id, scopes, perspective_key) where upper_inf(recorded) do nothing
	`);
	return true;
};

// Derive observations from a user's graph and write the significant ones back, the
// derive_insights twin, returning how many were written.
export const deriveInsights = async (
	scopes?: string[],
	reflect: Reporter<InsightReport> = report
): Promise<number> => {
	const key = scopeKey(scopes);
	const grounding = await asSystem(key, async (tx) => {
		const rows = (await tx.execute(sql`
			select statement from live_fact where predicate != ${OBSERVES}
			order by lower(recorded) desc limit ${settings.insightFactsK}
		`)) as unknown as { statement: string }[];
		return rows.map((row) => row.statement);
	});
	if (grounding.length < 2) return 0;
	const insight = await reflect(
		settings.insightSystem,
		'Facts:\n' + grounding.map((statement) => `- ${statement}`).join('\n')
	);
	const kept = keptObservations(insight);
	if (!kept.length) return 0;
	const vectors = await embed(kept.map((obs) => obs.statement), 'document');
	const nodeId = entityId(OBSERVATION_NODE, OBSERVATION);
	return asSystem(key, async (tx) => {
		await tx.execute(sql`
			insert into entity_content (id, name, type)
			values (${nodeId}, ${OBSERVATION_NODE}, ${OBSERVATION})
			on conflict do nothing
		`);
		await tx.execute(sql`
			insert into entity_claim (id, content_id, created_by, scopes)
			values (uuidv7(), ${nodeId}, ${settings.systemUserId}, ${uuidArray(key)}::uuid[])
			on conflict (content_id, scopes) do nothing
		`);
		let written = 0;
		for (const [index, obs] of kept.entries())
			written += (await writeObservation(tx, key, nodeId, obs, vectors[index])) ? 1 : 0;
		return written;
	});
};
